common-app-lib/src/cmr/common_app/app_events.clj

```diff
@@ -0,0 +1,74 @@
+(ns cmr.common-app.app-events
+  "Common events which applications can use when starting and stopping."
+  (:require
+   [clojure.java.io :as io]
+   [cmr.common-app.config :as config]
+   [cmr.common.log :refer [debugf infof warnf errorf]]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(import 'com.sun.management.HotSpotDiagnosticMXBean)
+(import 'java.lang.management.ManagementFactory)
+
+;;
+;; The internet suggests the following as how one would write the dump file in java:
+;;
+;;public static void dumpHeap (String filePath, boolean live) throws IOException {
+;;  MBeanServer server = ManagementFactory.getPlatformMBeanServer ();
+;;  HotSpotDiagnosticMXBean mxBean = ManagementFactory.newPlatformMXBeanProxy (server,
+;;      "com.sun.management:type=HotSpotDiagnostic", HotSpotDiagnosticMXBean.class);
+;;   mxBean.dumpHeap (filePath, live);
+;;}
+
+(defn- dump-heap
+  "Write a diagnostics file to the file system for analysis in Visual VM"
+  [^String app-name ^String file-path ^java.lang.Boolean live]
+  (let [bean (-> (ManagementFactory/getPlatformMBeanServer)
+                 (ManagementFactory/newPlatformMXBeanProxy "com.sun.management:type=HotSpotDiagnostic"
+                                                           HotSpotDiagnosticMXBean))]
+    (try
+      (when (.exists (io/file file-path))
+        (warnf (format "dump-heap: %s deleting diagnostic file %s." app-name file-path))
+        (io/delete-file file-path))
+      (.dumpHeap bean file-path live)
+      (infof "dump-heap: %s finished creating diagnostic file %s." app-name file-path)
+      (catch Exception err
+        (errorf "dump-heap: %s because %s" app-name (.getMessage err))))))
```

Comment:
Does the dump-heap take long to complete? Any chance of it getting "stuck" so that the shutdown hangs?

Reply:
it has not taken long for me, but it could hang sure. I think that AWS would just kill more aggressively in that case. I could time it if you like and write the time out.

Reply:
Capturing the time might be interesting yeah, thanks. And yeah, I'd expect that AWS would kill the task if shutdown hung for any reason, including this.